File: lib/ts/uapi/internal/ProcessRequestObject.ts
import { Message } from 'uapi/Message';
import { Serializer } from 'uapi/Serializer';
import { UApiError } from 'uapi/UApiError';
import { timeout } from 'util';

export async function processRequestObject(
    requestMessage: Message,
    adapter: (message: Message, serializer: Serializer) => Promise<Message>,
    serializer: Serializer,
    timeoutMsDefault: number,
    useBinaryDefault: boolean,
): Promise<Message> {
    const header: Record<string, any> = requestMessage.header;

    try {
        if (!header.hasOwnProperty('tim_')) {
            header['tim_'] = timeoutMsDefault;
        }

        if (useBinaryDefault) {
            header['_binary'] = true;
        }

        const timeoutMs = header['tim_'] as number;

        const responseMessage = await timeout(timeoutMs / 1000, adapter(requestMessage, serializer));

        if (responseMessage.body === { ErrorParseFailure_: { reasons: [{ IncompatibleBinaryEncoding: {} }] } }) {
            header['_binary'] = true;
            header['_forceSendJson'] = true;

            return await timeout(timeoutMs / 1000, adapter(requestMessage, serializer));
        }

        return responseMessage;
    } catch (e) {
        throw new UApiError();
    }
}

File: lib/java/src/main/java/uapi/internal/ProcessRequestObject.java
package uapi.internal;

import java.util.List;
import java.util.Map;
import java.util.Objects;
import java.util.concurrent.Future;
import java.util.concurrent.TimeUnit;
import java.util.function.BiFunction;

import uapi.Message;
import uapi.Serializer;
import uapi.UApiError;

public class ProcessRequestObject {
    public static Message processRequestObject(Message requestMessage,
            BiFunction<Message, Serializer, Future<Message>> adapter, Serializer serializer, long timeoutMsDefault,
            boolean useBinaryDefault) {
        final Map<String, Object> header = requestMessage.headers;

        try {
            if (!header.containsKey("time_")) {
                header.put("time_", timeoutMsDefault);
            }

            if (useBinaryDefault) {
                header.put("_binary", true);
            }

            final var timeoutMs = ((Number) header.get("time_")).longValue();

            final var responseMessage = adapter.apply(requestMessage, serializer).get(timeoutMs, TimeUnit.MILLISECONDS);

            if (Objects.equals(responseMessage.body,
                    Map.of("ErrorParseFailure_",
                            Map.of("reasons", List.of(Map.of("IncompatibleBinaryEncoding", Map.of())))))) {
                // Try again, but as json
                header.put("_binary", true);
                header.put("_forceSendJson", true);

                return adapter.apply(requestMessage, serializer).get(timeoutMs,
                        TimeUnit.MILLISECONDS);
            }

            return responseMessage;
        } catch (Exception e) {
            throw new UApiError(e);
        }
    }
}
